### PR TITLE
refactor: Move end-to-end test server fixture into its own module

### PR DESCRIPTION
**Pull request by alamb**
Re: https://github.com/influxdata/influxdb_iox/issues/890

# Rationale:

This is the first in what I hope to be a series of refactors to better manage the end to end tests. Specifically I want to be able to control and report on the tests individually as a normal rust test (each be a `#[tokio::test]`) as well as run them in parallel

If we like the pattern established in this PR, I will port the rest of the tests so they run as their own individual tests and make the harness more sophisticated (e.g. so setting writer_id doesn't trample on each other) but I didn't want to do so until I had some agreement this was what was an improvement.

# Changes:

1. Pulls the creation / setup of the server under test into a global lazily initialized fixture so that multiple tests can use it but only one server is started
2. Moves the write_api test into its own test as a demonstration of the technique working
3. Route logging to a log file into a temp file (filename printed to stdout)

Regarding logging: it would be nice to just print all the child's output to stdout but I couldn't figure that out with some quick googling and figured this would be ok for now.

# Example use:
Basic output (showing `test_write` running as its own test)
```
cargo test --test end-to-end

running 2 tests
test end_to_end_cases::write_api::test_write ... ok
test read_and_write_data ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

With more verbosity:

```
cargo test --test end-to-end -- write_api --nocapture

running 1 test
****************
Server Logging to "/var/folders/s3/h5hgj43j0bv83shtmz_t_w400000gn/T/server_fixture.log"
****************
Waiting for HTTP server to be up: error sending request for url (http://127.0.0.1:8090/health): error trying to connect: tcp connect error: Connection refused (os error 61)
Waiting for gRPC API to be up: Connection error: transport error: error trying to connect: tcp connect error: Connection refused (os error 61)
Waiting for HTTP server to be up: error sending request for url (http://127.0.0.1:8090/health): error trying to connect: tcp connect error: Connection refused (os error 61)
Waiting for gRPC API to be up: Connection error: transport error: error trying to connect: tcp connect error: Connection refused (os error 61)
Successfully connected to server
Successfully got a response from HTTP: Response { url: Url { scheme: "http", username: "", password: None, host: Some(Ipv4(127.0.0.1)), port: Some(8090), path: "/health", query: None, fragment: None }, status: 200, headers: {"x-powered-by": "Routerify v2.0.0-beta-2", "content-length": "2", "date": "Tue, 09 Mar 2021 13:12:18 GMT"} }
Storage service is running
Server is up correctly
[tests/end_to_end_cases/write_api.rs:58] err = ServerError(
    Status {
        code: InvalidArgument,
        message: "Violation for field \"lp_data\": Invalid Line Protocol: A generic parsing error occurred: TakeWhile1",
        details: b"\x08\x03\x12bViolation for field \"lp_data\": Invalid Line Protocol: A generic parsing error occurred: TakeWhile1\x1a}\n)type.googleapis.com/google.rpc.BadRequest\x12P\nN\n\x07lp_data\x12CInvalid Line Protocol: A generic parsing error occurred: TakeWhile1",
        metadata: MetadataMap {
            headers: {
                "content-type": "application/grpc",
                "date": "Tue, 09 Mar 2021 13:12:18 GMT",
            },
        },
    },
)
[tests/end_to_end_cases/write_api.rs:70] err = ServerError(
    Status {
        code: NotFound,
        message: "Resource database/Non_existent_database not found",
        details: b"\x08\x05\x121Resource database/Non_existent_database not found\x1aP\n+type.googleapis.com/google.rpc.ResourceInfo\x12!\n\x08database\x12\x15Non_existent_database",
        metadata: MetadataMap {
            headers: {
                "content-type": "application/grpc",
                "date": "Tue, 09 Mar 2021 13:12:18 GMT",
            },
        },
    },
)
test end_to_end_cases::write_api::test_write ... ok
```

```
alamb@ip-10-0-0-124 influxdb_iox2 % cat /var/folders/s3/h5hgj43j0bv83shtmz_t_w400000gn/T/server_fixture.log
cat /var/folders/s3/h5hgj43j0bv83shtmz_t_w400000gn/T/server_fixture.log
Mar 09 08:12:17.638  WARN influxdb_iox::influxdb_ioxd: NO PERSISTENCE: using Memory for object storage
Mar 09 08:12:17.639  INFO influxdb_iox::influxdb_ioxd: gRPC server listening bind_address=127.0.0.1:8092
Mar 09 08:12:17.650  INFO influxdb_iox::influxdb_ioxd: HTTP server listening bind_address=127.0.0.1:8090
Mar 09 08:12:17.650  INFO influxdb_iox::influxdb_ioxd: InfluxDB IOx server ready git_hash="8927b175f2ebe7a9e66c82551b951408ca411afc"
Mar 09 08:12:18.122  INFO influxdb_iox::influxdb_ioxd::http: Processing request request=Request { method: GET, uri: /health, version: HTTP/1.1, headers: {"accept": "*/*", "host": "127.0.0.1:8090"}, body: Body(Empty) }
Mar 09 08:12:18.122  INFO influxdb_iox::influxdb_ioxd::http: Successfully processed request response=Response { status: 200, version: HTTP/1.1, headers: {"x-powered-by": "Routerify v2.0.0-beta-2"}, body: Body(Full(b"OK")) }
```

# Notes:

I would basically plan the following next steps:
1.  Use a different server process (different ports) for testing setting writer id so that doesn't interfere with other tests
2. Port over the rest of the tests to use the fixture.

- [x] I've read the contributing section of the project [README](https://github.com/influxdata/influxdb_iox/blob/main/README.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

### CODE REVIEW

File: tests/end-to-end.rs [hunk 1]
```diff
@@ -345,118 +319,3 @@ fn substitute_nanos(ns_since_epoch: i64, lines: &[&str]) -> Vec<String> {
         })
         .collect()
 }
-
```
Comment (alamb) [inline tests/end-to-end.rs:348]: This got moved into its own module -- the actual logic is the same

File: tests/end-to-end.rs [hunk 2]
```diff
@@ -104,7 +78,7 @@ async fn read_and_write_data() {
     .await;
     management_api::test(&mut management_client).await;
     management_cli::test(GRPC_URL_BASE).await;
-    write_api::test(grpc).await;
+    //write_api::test(grpc).await;
```
Comment (alamb) [inline tests/end-to-end.rs:81]:
```suggestion
```

This no longer has to be called from `end-to-end.rs` , rather it now gets collected by the test runner

File: tests/common/server_fixture.rs [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+struct TestServer {
+    /// Is the server ready to accept connections?
+    ready: Mutex<bool>,
+
+    server_process: Child,
+
+    // The temporary directory **must** be last so that it is
+    // dropped after the database closes.
+    #[allow(dead_code)]
+    dir: TempDir,
```
Comment (tustvold) [inline tests/common/server_fixture.rs:98]: I think you can call this _dir and remove the #[allow]

Comment (alamb) [reply tests/common/server_fixture.rs:98]: Good idea -- this was just copied / pasted but I can clean it up as part of this PR

File: tests/common/server_fixture.rs [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+            //.arg("-vv")
+            .env("INFLUXDB_IOX_DB_DIR", self.dir.path())
+            .env("INFLUXDB_IOX_ID", "1")
+            .spawn()
+            .unwrap();
+        Ok(())
+    }
+
+    async fn wait_until_ready(&self) {
+        let mut ready = self.ready.lock().await;
```
[the remark this answers is not in the document]
Comment (alamb) [reply tests/common/server_fixture.rs:157]: In b86c21e77f60b9b3a8eea5a74938fa3505e7e1b4

File: tests/common/server_fixture.rs (hunk 3 above)
Comment (alamb) [reply tests/common/server_fixture.rs:98]: In b86c21e77f60b9b3a8eea5a74938fa3505e7e1b4

**Review by tustvold — APPROVED** [review]

**Comment by alamb** [comment]
🎉

**Pull request merged**